Extension lifecycle framework proposal (#601)
ghc-proposals/ghc-proposals: proposals/0000-extension-lifecycle-framework.md

 
 * `Legacy` extensions are explicitly not recommended for new code, much like `Deprecated`, however, they are expected to be supported indefinitely. This may be to maintain an older language set, maintain a bridge to a newer feature, or otherwise keep backward compatibility. Expected legacy extensions include `CUSKs`, `DeepSubsumption`,  and `NPlusKPatterns`.

Above we list expected categorizations of several extensions based on statements in the user's guide. We do not attempt to prescribe any particular categorizations as part of this proposal and simply provide a small number of expectations from the perspective of someone reading the existing documentation.

> The expected extension lifecycle includes the following transitions:

Reviewer: Do we want to say something like "A Mature extension should either eventually become part of a GHCXXXX language set or be deprecated"?

Author: I don't think so. It seems unlikely to me that `MultiWayIf` will ever be either deprecated or part of a default language set, for example. This proposal is not about changing the set of available extensions, it's about more clearly and accessibly communicating about those that exist.

That could be an interesting proposal, but it's not what we're proposing here. Should it be?

Reviewer 2: Some think that such a policy is desirable, but not everybody. Arnaud is in the process of shedding some light into this, but I think we can discuss it independent of the present proposal of warning against experimental extensions.

Reviewer: I guess the stated scope of this proposal seems somewhat broad, so I wonder if we want to talk about other aspects of the lifecycle. We can always do that later. And once we have these people will start asking questions "how long does something have to be in Deprecated before it gets removed?" etc.

Reviewer 3: @michaelpj The language was very careful to not specify a global deprecation time. Instead we left that on purpose to be determined at the time of deprecation for each extension. The intent is to require clear communication for a given extension but also is not so strict as to negatively impact GHC development.

Author: I actually don't think we should have a specific policy - language extensions are frequently part of a research program, and research can by definition fail! Research usually also has uncertain timelines, and part of evaluating work on language design is seeing how it is used in practice. It would be unfortunate for a strict policy to get in the way of that.

The lifecycle system proposed here is intended primarily to present facts about the extensions as they are at the time of a given release. The reason why we would like to propose that "deprecated" extensions have _some_ deadline for their removal is simply that we today really have two classes of extensions described as deprecated: those for which there is a real intention to remove them, and those that are considered a bad idea but may stick around forever. If I have a large, long-lived Haskell codebase, I need to react differently to the former warning than to the latter, as the former indicates that my existing code has an expiration date, while the latter is probably mostly important as a way to keep junior developers from introducing things that we would rather avoid if they've been reading old resources. It seems valuable to distinguish between these two states.

In other words, we're really proposing a simple but still informative and actionable model for describing language extensions WRT their likelihood of introducing bit-rot, along with a low-impact and inexpensive way for the compiler to help people act based on this information, not to impose a policy on all future R&D.

Perhaps the proposal should be amended to make it more clear that "expected" here is not normative, but rather descriptive of what we believe is the most likely path? Maybe the text just after the transitions should be before them?

Author: (BTW: I just emailed Arnaud about coordinating our work)

Author: For posterity: Arnaud got back to me quickly and indicated that he agrees that his work is largely orthogonal.

Author: (As he wrote below, but these threads can be hard to follow without a bit of redundancy)

Author: I've now clarified that the expected transitions are non-normative.

 * `Experimental` -> `Mature`
 * `Experimental` -> `Deprecated`
 * `Mature` -> `Legacy`
 * `Mature` -> `Deprecated`
 * `Legacy` -> `Deprecated`
 * `Deprecated` -> `Legacy`

However, it also seems plausible that new knowledge might from time to time cause a mature extension to once again be considered experimental, e.g. in the face of soundness bugs or subtle interactions with other features. We also do not rule out any transition explicity to allow for unforeseen circumstances. 

> For existing, or future, language sets such as `GHC2021` or `Haskell98`, it is expected that none of the contained extensions would be `Experimental`.

Author: Haskell 98 certainly includes `NPlusKPatterns`, which I think may end up in one of those categories. And over time, it doesn't seem implausible that we'd declare legacy an extension that had been in one of these default sets (after a very, very long notice period, of course)

> For existing, or future, language sets such as `GHC2021` or `Haskell98`, it is expected that none of the contained extensions would be `Experimental`.

Reviewer: Right, things that are in a set can become deprecated or legacy, but when we create a set we probably shouldn't include anything in those states? e.g. GHC2021 does _not_ contain `NPlusKPatterns`

> For existing, or future, language sets such as `GHC2021` or `Haskell98`, it is expected that none of the contained extensions would be `Experimental`.

Reviewer 3: It certainly seems like a good idea to as a matter of policy say that a new language set does not contain legacy or deprecated extensions. However, if a language set should be added after the fact for some reason, then perhaps it would include a legacy extension. Maybe resolving if a policy should exist for this and what exactly that would be is worthy of a separate discussion?

> For existing, or future, language sets such as `GHC2021` or `Haskell98`, it is expected that none of the contained extensions would be `Experimental`.

Author: I'm loathe to impose policy here. I think that the people building such a set can take these extensions into account, and that any policy is either redundant or the kind of thing that would be set aside anyway in the situation.

> For existing, or future, language sets such as `GHC2021` or `Haskell98`, it is expected that none of the contained extensions would be `Experimental`.

Author: One more case: the `Haskell98` extension set presumably came into existence after some of its component parts were broadly acknowledged as design dead-ends (and thus `Legacy`). Introducing this set is nonetheless useful, as it helps people improve compatibility with older code. Here, any policy could have been set aside, but making a policy seems to not buy us anything. I certainly agree that we should be surprised if e.g. `GHC2035` includes any deprecated extensions when it's introduced in 2035, but how will a written policy actually help?

### 2.2. User Interface

Haskell users have different tolerances for risks related to language change. Some derive great value from using the newest features, while others are conservative. GHC should serve the whole spectrum.

There will be an additional set of warnings:
 * `-WXDeprecated`: Issue a warning when a deprecated extension is used

>  * `-WXMature`: Issue a warning when a mature extension is used

Author: Not on its own. It seems plausible that it might be useful together with some more specific flags to configure a policy, but it's mostly there for the sake of consistency and completeness.

>  * `-WXMature`: Issue a warning when a mature extension is used

Reviewer 4: Assuming it means "when a mature extension is mentioned in a `LANGUAGE` pragma" then I can imagine it being useful for identifying modules with `LANGUAGE` pragmas that are redundant given the global settings.

>  * `-WXMature`: Issue a warning when a mature extension is used

Author: This is not what it means - but I think we addressed this later on in the discussion.

 * `-WXExperimental`: Issue a warning when an experimental extension is used 
 * `-WXLegacy`: Issue a warning when a legacy extension is used.

Each category will also support the usual`-Wno-` syntax, so `-Wno-XDeprecated` will turn off warnings for deprecated extensions. Additionally, each extension should have a configurable warning that can be individually enabled or disabled, allowing users and teams to inform GHC about their local policies and needs. For instance, `-WXDeprecated -Wno-XRank2Types` would warn about all deprecated extensions except `Rank2Types`.

> Furthermore, `-WXDeprecated` and `-WXExperimental` would be added to `-Wcompat` for the next release to allow the community time to adjust prior to being on by default. However, the goal is for `-WXDeprecated` and `-WXExperimental` to be on by default.

Author: That's definitely worth considering. The idea here was to warn people away from extensions that are likely to change incompatibly, while the definition of `Legacy` is that they are expected to be stable, but not recommended. Perhaps this should be in `-Wall`?

> Furthermore, `-WXDeprecated` and `-WXExperimental` would be added to `-Wcompat` for the next release to allow the community time to adjust prior to being on by default. However, the goal is for `-WXDeprecated` and `-WXExperimental` to be on by default.

Reviewer: I agree that it's less bad than deprecated or experimental stuff - those have the connotation that you might be broken (by removal or change), whereas legacy is just maybe not a good idea.

> Furthermore, `-WXDeprecated` and `-WXExperimental` would be added to `-Wcompat` for the next release to allow the community time to adjust prior to being on by default. However, the goal is for `-WXDeprecated` and `-WXExperimental` to be on by default.

Author: That was the thought process - our primary goal is to help teams that include junior developers more easily figure out which features to use. The organizational risk of an extension that will be deleted in three releases is much higher than the risk posed by an extension that's got a funny odor and may not work well with other features, but will be kept around indefinitely.

## 4. Effect and Interactions

The tension between large-scale industrial users of Haskell who desire stability and predictability and GHC's traditional role as a venue for language experiments is very real. This system addresses some of the industrial concern by providing clear communication to users that sets realistic expectations for language extensions. By providing users with an explicit way to indicate their own tolerances for legacy, experimental, and deprecated features, this proposal aims to ease the tension in a way that requires as little as possible additional investment on the part of the GHC team.

The primary interaction that we anticipate from this change is with existing warning flags and build configurations. Because this proposal does not change the semantics of any programs, we expect that it will not be a major disturbance.

## 5. Costs and Drawbacks

Development and ongoing costs are expected to exist primarily in the form of additional time to sufficiently warn users of upcoming changes, creating and updating the status of extensions, and maintenance of the warning flags.

> Organizations that use `-Werror` may incur a one-time cost in updating their compiler flags to disable experimental or deprecated extension warnings.

Author: That was what we intended by "updating their compiler flags to disable ... warnings" - this does have a nonzero one-time cost that I think we should acknowledge.

> Organizations that use `-Werror` may incur a one-time cost in updating their compiler flags to disable experimental or deprecated extension warnings.

Reviewer: Ah I see, I read this wrongly as "updating their code to disable experimental or deprecated extensions", which is not actually what it says.

> Organizations that use `-Werror` may incur a one-time cost in updating their compiler flags to disable experimental or deprecated extension warnings.

Author: Do you think others are likely to have the same understanding? If so, we should modify the text to make it more clear.

> Organizations that use `-Werror` may incur a one-time cost in updating their compiler flags to disable experimental or deprecated extension warnings.

Author: Done

The weakness of this model is that, over time, the accumulation of programming extensions in various states of use and disuse can create confusion, especially for developers being introduced to Haskell. There is not one Haskell, but many, and each individual or team must necessarily choose their own. Not everyone has yet had the opportunity to accumulate the experience and knowledge that will allow them to quickly determine whether or not to use a given language extension, leading to various points of stress.

The authors of this proposal believe that a clear lifecycle for these programming language extensions would make it easier for newcomers, those responsible for onboarding new Haskell developers as well as managers, and the community more broadly, to make sense of the Haskell language extensions. This framework would make it clear which extensions are prone to rapid evolution, which of them are stable and suitable for serious use, and which of them are generally not considered suitable for new development.

## 2. Proposed Change Specification

### 2.1. Categories and Lifecycle
Language extensions will be classified into the following categories

>  * `Stable` extensions are considered to be finished and are not expected to undergo regular changes. These features can be used without worry of unexpected changes, and they are not known to contain serious design or implementation deficiencies. Any breaking change to a stable extension will be announced well in advance of the change being made, with a migration path provided if possible. For an extension to be classified as `Stable` it must be considered `Stable` when used in combination of all possible, non-mutually exclusive, extensions that are already `Stable`. This ensures any extension that is `Stable` will be have the same use without worry when combined with any other extension in the set, while allowing for mutually exclusive extensions to be added. Ideally, no breaking change will be made to a `Stable` extension, with incompatible changes resulting instead in a new, related extension, that is possibly mutually exclusive with the existing one, to enable smooth migration. Some extensions in this category might be `MultiWayIf`, `MonoLocalBinds`, and `ViewPatterns`.

Reviewer 5: ```suggestion
 * `Stable` extensions are considered to be finished and are not expected to undergo regular changes. These features can be used without worry of unexpected changes, and they are not known to contain serious design or implementation deficiencies.   More specifically
  * For an extension to be classified as `Stable` it must be considered `Stable` when used in combination of all possible, non-mutually exclusive, extensions that are already `Stable`. This ensures any extension that is `Stable` will be have the same use without worry when combined with any other extension in the set, while allowing for mutually exclusive extensions to be added. 
  * Ideally, no breaking change will be made to a `Stable` extension.  Instead, it may be better to add a new, related extension, that is possibly mutually exclusive with the existing one, to enable smooth migration. 
  * If, however, a breaking change is made to a stable extension, it will be announced well in advance of the change being made, with a migration path provided if possible.   
  * Some extensions in the `Stable` category might be `MultiWayIf`, `MonoLocalBinds`, and `ViewPatterns`.
```

Reviewer 5: "This ensures any extension that is `Stable` will be have the same use without worry " doe not parse.

Reviewer 5: Any thoughts on this?

Reviewer 3: I thought the wording changed slightly in the last update but I must have put the existing one back before committing.
Perhaps:

> This ensures any extension that is `Stable` will have the same stability properties when combined with any other extension in the set, while allowing for mutually exclusive extensions to be added.
?

>  * `Experimental` extensions are undergoing active development, or have consequences that make it impossible in practice to avoid breaking changes. The syntax and semantics that are enabled by the extension are likely to change regularly. It is expected that most new language extensions will begin as experimental, while the developers and the community gain experience with their use. Despite being open to breaking change, an `Experimental` extension must be `Deprecated` prior to removal. Some extensions in this category might be `RequiredTypeArguments`, `RebindableSyntax`, and `Arrows`.

Reviewer 5: ```suggestion
 * `Experimental` extensions are undergoing active development, or have consequences that make it impossible in practice to avoid breaking changes. The syntax and semantics that are enabled by the extension are likely to change regularly. 
 * It is expected that most new language extensions will begin as experimental, while the developers and the community gain experience with their use. 
 * Despite being open to breaking change, an `Experimental` extension must be `Deprecated` prior to removal. 
 * Some extensions in this category might be `RequiredTypeArguments`, `RebindableSyntax`, and `Arrows`.
```

>  * `Deprecated` extensions are considered to be design or implementation dead ends, and should not be used in new code. Deprecating an extension means that GHC intends to remove support for it in a future release (though this decision may be revised in the light of user feedback). When an extension is deprecated, the user's guide and the compiler warning _must_ include a statement about the longevity of the extension, though this need not necessarily commit to a concrete time for removal. Additionally, the user's guide and the warning should direct users to information about migrating away from the deprecated extension. Any extension will be deprecated prior to removal. Some extensions in this category might be `OverlappingInstances`, `Rank2Types`, `RecordPuns`, and `TypeInType`.

Reviewer 5: ```suggestion
 * `Deprecated` extensions are considered to be design or implementation dead ends, and should not be used in new code. Deprecating an extension means that GHC intends to remove support for it in a future release (though this decision may be revised in the light of user feedback). 
   * When an extension is deprecated, the user's guide and the compiler warning _must_ include a statement about the longevity of the extension, though this need not necessarily commit to a concrete time for removal. Additionally, the user's guide and the warning should direct users to information about migrating away from the deprecated extension. 
   * Any extension will be deprecated prior to removal.
   * Some extensions in the `Deprecated` category might be `OverlappingInstances`, `Rank2Types`, `RecordPuns`, and `TypeInType`.
```

Reviewer 3: Thank you, the last update should fix the formatting issue.

 * `Legacy` extensions are explicitly not recommended for new code, much like `Deprecated`, however, they are expected to be supported indefinitely. This may be to maintain an older language set, maintain a bridge to a newer feature, or otherwise keep backward compatibility. Extensions that are `Legacy` may have some conflict with `Stable` extensions and explicitly do not need to be considered for an extension to be `Stable`. Some extensions in this category might be `CUSKs`, `DeepSubsumption`,  and `NPlusKPatterns`.

Above we list expected categorizations of several extensions based on statements in the user's guide. We do not attempt to prescribe any particular categorizations as part of this proposal and simply provide a small number of expectations from the perspective of someone reading the existing documentation.

> The above places some restrictions on the transitions included in the extension lifecycle. The following transitions are also expected to be included:

Reviewer 5: Please list all transitions.  Otherwise what does "also" mean, precisely?  

Or, perhaps better, list the *excluded* transitions, with some rationale.

Reviewer 5: I am keen to list all allowed transitions; or all disallowed transitions.  Listing some of each, leaving others unspecified is not a happy state.

Reviewer 3: Updated to list all transitions, including those not explicitly included or excluded in the proposal.

 * `Experimental` -> `Stable`
 * `Experimental` -> `Deprecated`
 * `Experimental` -> `Legacy`
 * `Stable` -> `Legacy`
 * `Legacy` -> `Deprecated`

> However, it also seems plausible that new knowledge might from time to time cause a stable extension to once again be considered experimental, e.g. in the face of soundness bugs or subtle interactions with other features. We explicitly allow for more transitions to allow for unforeseen circumstances. Restrictions are places on removal only. With this we aim to balance predictibility and flexibility. The consideration of both is important as extension classifications are first and foremost a communications channel between language designers, compiler developers, and users.

Reviewer 5: "  Restrictions are places on removal only"

What precisely is the restriction?.

Reviewer 5: Any thoughts on this?

Reviewer 3: I updated the set of included transitions to be more complete and added the exclusion list.

> However, it also seems plausible that new knowledge might from time to time cause a stable extension to once again be considered experimental, e.g. in the face of soundness bugs or subtle interactions with other features. We explicitly allow for more transitions to allow for unforeseen circumstances. Restrictions are places on removal only. With this we aim to balance predictibility and flexibility. The consideration of both is important as extension classifications are first and foremost a communications channel between language designers, compiler developers, and users.

Reviewer 5: " We explicitly allow for more transitions".  More than what?   Can you make this precise?

Maybe you have in mind a set of "normal" transitions, ones that are "exceptional" and ones that are "disallowed"?  I'm really not sure.

Reviewer 3: Updated For this and the above to note the excluded transitions explicitly.

Reviewer 5: I can't see the change here.

Reviewer 3: The explicitly excluded transitions are listed in the block right above this paragraph now.

> For existing, or future, language sets such as `GHC2021` or `Haskell98`, it is expected that none of the contained extensions would be `Experimental`. However, this proposal does not seek to impose any particular policy on the inclusion of extensions into language sets - the developers and the steering committee are always in the best position to make a decision about a concrete extension and extension set.

Reviewer 5: "language edition" not "language set"

Reviewer 3: Thank you, updated!

Reviewer 5: It still says "language set"

Moving an extension through the lifecycle will require a GHC proposal. This will give users a chance to provide their input, it will help catch unexpected interactions (such as an `Stable` extension implying an `Experimental` one, and thus potentially allowing breakage at any  moment), and it provides a way for language implementers and users to clarify their mutual expectations.

### 2.2. Documentation

> The user's guide for each extension documents where it is in the extension lifecycle. We expect this to be next to the "Since" and "Implied by" fields in extension documentation. Additionally, the user's guide should provide a history of the extension's sojourn through the various states, and the GHC versions in which it went from being experimental to stable, or from stable to deprecated.

Reviewer 5: Why do you pull out two particular transitions for special mention?  Why not just say the user guide should document the GHC version in which each transition was made?

Reviewer 3: I've updated the wording to make it more clear the guide should say when each transition was made and that two called out are examples.

Reviewer 5: I don't see a change here

The authors of this proposal believe that a clear lifecycle for these programming language extensions would make it easier for newcomers, those responsible for onboarding new Haskell developers, and the community more broadly, to make sense of the Haskell language extensions. This framework would make it clear which extensions are experimental and prone to rapid evolution, which of them are mature and thus relatively stable and suitable for serious use, and those that are generally not considered suitable for new development. Extending GHC's warnings to account for these categories allows projects to find out when features they use have been deprecated in a new compiler version as well as to specify general policies about language stability that can be checked mechanically.

## 2. Proposed Change Specification

### 2.1. Categories and Lifecycle
Language extensions will be classified into the following categories

>  * `Experimental` extensions are undergoing active development. The syntax and semantics that are enabled by the extension are likely to change regularly. It is expected that most new language extensions will begin as experimental. At the time of writing, `LinearTypes` and `OverloadedRecordUpdate` seem to be in this category.

Reviewer 6: `Experimental` covers two possibly non-overlapping categories: newer extensions and extensions that are likely to change. I definitely want to be warned about the latter, but not the former. Perhaps another category is in order: `Nascent`

Reviewer 5: > Experimental covers two possibly non-overlapping categories: newer extensions and extensions that are likely to change. I definitely want to be warned about the latter, but not the former. Perhaps another category is in order: Nascent

Let's not have too many categories!  It would just get confusing and harder to decide which category to put something into.  Newer extensions are, because they are new, quite likely to change.  Less is more :-)

Perhaps it would help to frame the categories in actionable form:

* `Stable` (which I agree is better than `Mature`): you can rely on this; it's unlikely to change.
* `Experimental`: quite likely to change.
* `Deprecated`: try to avoid using this at all, and actively remove it from code you maintain.
* `Legacy`: avoid for new code.

Reviewer 3: > > Experimental covers two possibly non-overlapping categories: newer extensions and extensions that are likely to change. I definitely want to be warned about the latter, but not the former. Perhaps another category is in order: Nascent
> 
> Let's not have too many categories! It would just get confusing and harder to decide which category to put something into. Newer extensions are, because they are new, quite likely to change. Less is more :-)

I agree with this completely. And if we learn more after living with this for a while, we can make adjustments!

> 
> Perhaps it would help to frame the categories in actionable form:
> 
> * `Stable` (which I agree is better than `Mature`): you can rely on this; it's unlikely to change.
`Stable` was specifically avoided based on prior feedback around language extension classification. That said, I personally agree that `Stable` is a better name, and would certainly endorse a change there.

> * `Deprecated`: try to avoid using this at all, and actively remove it from code you maintain.
And per the proposal, there is some statement about how long it will continue to exist.

> * `Legacy`: avoid for new code.
But it will continue to exist indefinitely.

Reviewer 4: > `Stable` was specifically avoided based on prior feedback around language extension classification.

I'm curious to know what the arguments are against `Stable`. From a brief skim of the discussion there seem to be quite a few people favouring it over `Mature`. And it seems to more clearly communicate "this isn't expected to change", whereas `Mature` implies "this has been around a while" without necessarily indicating that it is in a good state to use (at least to me). For example, I would say `FunctionalDependencies` is a very "mature" extension but would hesitate to declare it "stable" given the number of dark corners it exposes.

Reviewer 2: Isn't your last example precisely an argument for mature (implying sensible, reliable, not childish)? Stable only indicates ”unchanging”, but has otherwise no connotations of quality. All the legacy extensions are probably more stable than the mature ones…

To me, mature means “ready for production use” which _implies_ stable (or at least only backward compatible changes maybe), but says more actually.

But in the end it's just a name…

Reviewer 3: > > `Stable` was specifically avoided based on prior feedback around language extension classification.
> 
> I'm curious to know what the arguments are against `Stable`. From a brief skim of the discussion there seem to be quite a few people favouring it over `Mature`. And it seems to more clearly communicate "this isn't expected to change", whereas `Mature` implies "this has been around a while" without necessarily indicating that it is in a good state to use (at least to me). For example, I would say `FunctionalDependencies` is a very "mature" extension but would hesitate to declare it "stable" given the number of dark corners it exposes.

I _believe_ most of the arguments against were around how to precisely define what `Stable` means. 

> 
> To me, mature means “ready for production use” which _implies_ stable (or at least only backward compatible changes maybe), but says more actually.
> 
> But in the end it's just a name…

This, along with some notion that stable was not a good name, is exactly how mature came to be the name. To me whichever way makes the understanding the easiest to the most people is the way to go, which feels like `Stable` at the moment.

Reviewer 4: It's interesting how the same names can have different connotations to different people. :grin: I'm happy whichever way we end up painting this particular bikeshed.

Author: After consulting with my coauthors, I updated the proposal to call it `Unstable`. Thanks!

 
 * `Mature` extensions are considered to be finished and are not expected to undergo regular changes. These features can be used without worry of unexpected changes, and they are not known to contain serious design or implementation deficiencies. Any breaking change to a mature extension will be announced well in advance of the change being made, with a migration path provided if possible. Some extensions in this category might be `MultiWayIf`, `MonoLocalBinds`, and `ViewPatterns`.

 * `Deprecated` extensions are considered to be design or implementation dead ends, and should probably not be used in new code. Deprecating an extension _must_ come with a statement about the longevity of the extension. This need not be a commitment to remove it at a concrete time, but is a commitment to either remove it or transition it to another state at a time stated by GHC HQ when it is marked as `Deprecated`. Any extension will be deprecated prior to removal. Deprecated extensions are expected to include `IncoherentInstances`, `OverlappingInstances`, `Rank2Types`, `RecordPuns`, and `TypeInType`.
 

>  * `Legacy` extensions are explicitly not recommended for new code, much like `Deprecated`, however, they are expected to be supported indefinitely. This may be to maintain an older language set, maintain a bridge to a newer feature, or otherwise keep backward compatibility. Expected legacy extensions include `CUSKs`, `DeepSubsumption`,  and `NPlusKPatterns`.

Reviewer 4: I'm struggling to see the distinction between `Deprecated` and `Legacy`. It seems the only functional difference is that they belong to different warning groups, is that right?

Given that `Deprecated` extensions do not necessarily have a concrete schedule for removal, and there is nothing stopping a `Legacy` extension subsequently being re-categorised to `Deprecated` and later removed, are they really different? I'm not convinced that trying to guess which extensions are most likely to be removed in the future is really useful. Could we simplify the proposal and have a single category for deprecated extensions?

The examples don't really help me understand: there seems to be little motivation to remove synonyms like `Rank2Types` or `RecordPuns`, whereas I can imagine changes to the typechecker making it desirable to reduce maintenance costs by removing `DeepSubsumption` at some future point.

I suppose the distinction could be made more concrete if `Deprecated` extensions were always on track for removal in some relatively predictable number of releases. But it's pretty rare for GHC to actually remove extensions in my experience (and I think this is a good thing).

Reviewer 3: > I'm struggling to see the distinction between `Deprecated` and `Legacy`. It seems the only functional difference is that they belong to different warning groups, is that right?
> 

There is a huge difference when I've been working with people. The intent here is for users to get two different signals: 
- `Legacy`: We will continue to support this, but this is not recommended for new code. 
- `Deprecated`: This is marked for removal and is recommended to be purged from all codebases as soon as possible. If there is a compelling reason to keep it, bring that up before (timeframe set when extension was deprecated) which is the point at which it will be decided to actually remove or move to some other status.

In writing the proposal, care was taken to not be too prescriptive and concretely set a timeframe for all extensions to be removed when deprecating, as some exceptional case will certainly arise shortly after such strong rules are in place. However, we could always change the language to say the timeframe is for removal, not for the decision on removal, but that allows less community feedback on why something should be saved, no?

>  Could we simplify the proposal and have a single category for deprecated extensions?

I would strongly recommend against that. In fact I would say a huge part of the benefit of the proposal is in understanding what could actually be removed as opposed to what is not recommended but still explicitly supported going forward.

> The examples don't really help me understand: there seems to be little motivation to remove synonyms like `Rank2Types` or `RecordPuns`, whereas I can imagine changes to the typechecker making it desirable to reduce maintenance costs by removing `DeepSubsumption` at some future point.
> 
> I suppose the distinction could be made more concrete if `Deprecated` extensions were always on track for removal in some relatively predictable number of releases. But it's pretty rare for GHC to actually remove extensions in my experience (and I think this is a good thing).

Not removing extensions _ever_ actually has a large set of negative impacts. Firstly, when bringing someone to the language, we now get to say "learn Haskell" and oh learn this always increasing set of extensions that may or may not be on at any given moment when you're looking at a particular piece of code. Even if the impact to GHC is negligible for something like `Rank2Types`, the impact of having one less thing for a new user to see and trip over is actually important.  There are at least a handful of extensions that seem like good candidates to actually remove and with the warnings/notice in place from this proposal, I think that moving the forward with those deprecation and removal cycles would be beneficial to the ecosystem.

Reviewer 2: If we have doubts that we'll ever see a extension being Deprecated, I suggest to leave it out for now. Every concept imposes a mental load on our users, company policy makers will have to make decisions (e.g. how to treat deprecated extensions). That's a cost we should not incur prematurely (or is it postMaturely?) if it's possible that we won't use it. We can still introduce the concept when it becomes relevant.

Reviewer 4: Actually I wouldn't go so far as to say that we should *never* deprecate an extension, so I think it's helpful to retain the concept.  I just think we should do so very sparingly.

For example, I would say `IncoherentInstances` does meet the bar for deprecation and eventual removal, because it is basically always better to use per-instance pragmas. But I wouldn't remove `Rank2Types` because the cost of breaking code is greater than the cost of having an obsolete synonym for `RankNTypes` hanging around.

Reviewer 2: I am not saying that we should _never_ deprecate, and I agree we can retain the concept – but only actually introduce it to the actual code when we actually know for certain that we are going to deprecate an extension.

_If_ we know for sure that we will deprecate, say `IncoherentInstances`, then we can introduce `-WDeprecated` right away, of course. But if we are unsure about at this point, I’d advise against introducing a category that may potentially be empty for a long while.

Reviewer 4: Indeed I suppose it would be strange to have a warning `-WXDeprecated` that could never fire. But I guess we can resolve this when debating a concrete proposal for classification of extensions (which it sounds like will probably try to deprecate some).

Author: Even an unused `Deprecated` category has a purpose: it communicates that all these `Legacy` extensions (those marked "deprecated" in the manual today) are not in fact scheduled for removal, and that your code that uses them is not in an emergency state. Today, "deprecated" in software development (and much other Haskell practice) usually means "removal is on the way", but if users can rely on the alias `RecordPuns` sticking around essentially forever, then it's good for them to know that.

Thinking more about this, perhaps we should also write a quick sentence for each `Legacy` or `Deprecated` extension when doing the classification proposal that can be included in the warning. An extension might be considered to be a bad idea for technical reasons (`IncoherentInstances`), or just because it's been renamed to something else (`RecordPuns`). It would be nice if the warning could say `Rank2Types` is a synonym for `RankNTypes` in GHC and that `IncoherentInstances` should probably be replaced by pragmas.

Reviewer 2: Yes, absolutely: no proposal should ever be understood as forbidding better error messages! :-)

Maybe we should wait with “Do we actually need Deprecated or not” until after we have categorized all existing extensions (a daunting task anyways, worse than defining GHC20xx…).

Author: It sounds to me like there may be an actual plan to remove `TypeInType` - it's newly deprecated in 9.6.1. But we really do need to have a deliberate discussion about this in the follow-up.

Author: In any case, I'm marking this thread Resolved, as Adam's concrete formulations are in the current draft, and the authors will keep the `Deprecated` category.

Above we list expected categorizations of several extensions based on statements in the user's guide. We do not attempt to prescribe any particular categorizations as part of this proposal and simply provide a small number of expectations from the perspective of someone reading the existing documentation.

The expected extension lifecycle includes the following transitions:

>  * `Experimental` -> `Mature`
>  * `Experimental` -> `Deprecated`
>  * `Mature` -> `Legacy`
>  * `Mature` -> `Deprecated`
>  * `Legacy` -> `Deprecated`
>  * `Deprecated` -> `Legacy`

Reviewer 4: Given that we're trying to describe the "happy path" here, the following transitions seem logical to me:

```suggestion
 * `Experimental` -> `Mature`
 * `Experimental` -> `Legacy`
 * `Experimental` -> `Deprecated`
 * `Mature` -> `Legacy`
 * `Legacy` -> `Deprecated`
```

An experimental extension might move to any state. Mature extensions might fall out of favour but should typically not be deprecated/removed (at least without lingering in the legacy state for a good while).

Of course we might change course e.g. a deprecated extension might move back to legacy in the light of user feedback. But that seems adequately covered by the next paragraph?

Author: I think this is all addressed in the current draft. Thanks!

However, it also seems plausible that new knowledge might from time to time cause a mature extension to once again be considered experimental, e.g. in the face of soundness bugs or subtle interactions with other features. We also do not rule out any transition explicitly to allow for unforeseen circumstances. 

For existing, or future, language sets such as `GHC2021` or `Haskell98`, it is expected that none of the contained extensions would be `Experimental`. However, this proposal does not seek to impose any particular policy on the inclusion of extensions into language sets - the developers and the steering committee are always in the best position to make a decision about a concrete extension and extension set.

### 2.2. User Interface

Haskell users have different tolerances for risks related to language change. Some derive great value from using the newest features, while others are conservative. GHC should serve the whole spectrum.

There will be an additional set of warnings:

>  * `-WXDeprecated`: Issue a warning when a deprecated extension is used

Reviewer 4: Please can the proposal include a definition of what it means for an extension to be "used"? I suspect it means "when a `LANGUAGE` pragma in a module enables the extension", which is easily specified, and wouldn't warn if the extension was enabled in a `.cabal` file or on the command line. But it could also mean "when the feature enabled by the extension occurs in the program" (e.g. an occurrence of a string literal for `OverloadedStrings`).

Reviewer 2: I'd expected this to mean ”enabled”, independently of how (pragma, flag, cabal file).

I'd advise against differtiating between these ways to enable the extension, it would complicate the conceptual meta-theory of extensions. (Also, I am not sure if GHC even knows whether a flag was passed by the user, or by `cabal`).

```suggestion
 * `-WXDeprecated`: Issue a warning when a deprecated extension is enabled (by way of pragma, command like flag or otherwise)
```

Reviewer 4: Hmm, but then we have warnings that can arise just based on command-line options, entirely independently of source files. And that seems likely to lead to complications, because command-line options can also modify whether the warnings should be emitted, so it matters exactly when and where the check is done (cf. https://gitlab.haskell.org/ghc/ghc/-/issues/23402).

I can see that e.g. having `ghci -XSomeDeprecatedExtension` issue a warning makes sense, so I'm not strongly opposed to your suggestion. But the proposal should at least specify whether `ghc -WXDeprecated -XSomeDeprecatedExtension -Wno-XDeprecated` does or does not issue a warning.

Reviewer 2: Right, we should be clear about this.

My preference is to keep language pragmas and command line flags equivalent. This means that `-W` flags and `-X` flags commute, and `-X` flags ans pragmas have their effect when actually looking at the file.

Surely, we don’t want developers to think about having to write
```
{-# LANGUAGE SomeDeprecatedExtension #-}
{-# OPTIONS_GHC -Wno-XSomeDeprecatedExtension #-}
```
(which I expect will arise as a common idiom from all this) or
```
{-# OPTIONS_GHC -Wno-XSomeDeprecatedExtension #-}
{-# LANGUAGE SomeDeprecatedExtension #-}
```

I agree that there is existing complexity with warnings about flags, so there would be precedent. Maybe we can keep it contained to that dark corner :-)

 * `-WXMature`: Issue a warning when a mature extension is used
 * `-WXExperimental`: Issue a warning when an experimental extension is used 
 * `-WXLegacy`: Issue a warning when a legacy extension is used.

> Each category will also support the usual`-Wno-` syntax, so `-Wno-XDeprecated` will turn off warnings for deprecated extensions. Additionally, each extension should have a configurable warning that can be individually enabled or disabled, allowing users and teams to inform GHC about their local policies and needs. For instance, `-WXDeprecated -Wno-XRank2Types` would warn about all deprecated extensions except `Rank2Types`.

Reviewer: What exactly do the `-WXFoo` warnings do? Do they just generically warn about the extension, or do they specifically say something about it's status? e.g. does `-WXRank2Types` produce something like
```
WARNING: Use of deprecated extension Rank2Types
```
Is that still the case if I don't have `-WXDeprecated` and I just turn on `-WXRank2Types`?

The reason I ask is that if those flags provide generically worded warnings about the use of an extension, then this actually provides us with an additional feature: configurable sets of warnings for use of language extensions. This is useful in a team setting where you maybe want to decide that you don't want to use some extension. You can do this today with `hlint`, but this would provide a built-in way of doing it.

I think a message like this might straddle the gap nicely?
```
WARNING: Use of extension Rank2Types (Deprecated)
```
i.e. the category is given as additional information, not necessarily pointed to as the cause of the warning.

Author: This is a good point. Our intention when writing this was that they give a warning for using the extension (which could, of course, be converted into an error), but we didn't propose a specific wording. We did want teams to be able to use it in the way you're suggesting, with individual warnings for individual extensions in order to enforce team policies in CI.

Depending on what's reasonable to implement, I wonder whether the right thing to do is to say which flag actually enabled the extension in the warning message, and render it based on that. E.g.
```
Foo.hs:3:5: warning: [-WXRank2Types]
    Use of extension Rank2Types
```
```
Foo.hs:3:5: warning: [-WXDeprecated]
    The extension Rank2Types is deprecated
```
with the latter taking precedence:
```
Foo.hs:3:5: warning: [-WXDeprecated, -WXRank2Types]
    The extension Rank2Types is deprecated
```

This may be infeasible or too expensive though.

Reviewer 4: I think the square brackets should be consistent with the current behaviour. Currently GHC always shows the specific warning flag, plus the group(s) to which it belongs if `-fshow-warning-groups` is enabled. This is helpful because it is immediately obvious how to disable a warning selectively, even if it was enabled by a group. So I'd rather see something like:
```hs
Foo.hs:3:5: warning: [-WXRank2Types (in -WXDeprecated)]
    Use of extension Rank2Types, which is deprecated.

    This extension is an obsolete synonym for RankNTypes.

    Suggested fix: enable RankNTypes instead of Rank2Types.
```

```hs
Foo.hs:3:5: warning: [-WXGADTs (in -WXStable)]
    Use of extension GADTs, which is stable.
```

Author: Good point - I'll update the text after lunch to match your examples here.

Author: Done

---
author: Trevis Elser, Chris Dornan, David Thrane Christiansen
date-accepted: ""
ticket-url: ""
implemented: ""
---

This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/601).

# Extension Lifecycle Framework

> We propose a categorization scheme for Haskell language extensions. This scheme is simple, in that there are few categories that are described in terms of the user-relevant aspects.  The purpose of this proposal is not to classify all extensions - rather, we seek to establish a basis upon which to carry out that work subsequently. Example classifications are provided to illustrate the proposed framework, and acceptance of this proposal should not count as acceptance these examples.

Reviewer 3: Added this, thank you!

The weakness of this model is that, over time, the accumulation of programming extensions in various states of use and disuse can create confusion, especially for developers being introduced to Haskell. There is not one Haskell, but many, and each individual or team must necessarily choose their own. Not everyone has yet had the opportunity to accumulate the experience and knowledge that will allow them to quickly determine whether or not to use a given language extension, leading to various points of stress.

The authors of this proposal believe that a clear lifecycle for these programming language extensions would make it easier for newcomers, those responsible for onboarding new Haskell developers as well as managers, and the community more broadly, to make sense of the Haskell language extensions. This framework would make it clear which extensions are prone to rapid evolution, which of them are stable and suitable for serious use, and which of them are generally not considered suitable for new development.

## 2. Proposed Change Specification

### 2.1. Categories and Lifecycle

> Language extensions will be classified into the following categories

Reviewer 4: A small point, but what about classifying "negative" extensions? For example `FieldSelectors` was effectively `Stable` when introduced (whereas `NoFieldSelectors` was definitely `Experimental`).

I guess the simplest thing would be to say that we classify both the enabled and disabled states, but the normal case is for an extension to be `Stable` when disabled (with `NoFieldSelectors` being one of a small number of exceptions).

>  * `Stable` extensions are considered to be finished and are not expected to undergo regular changes. These features can be used without worry of unexpected changes, and they are not known to contain serious design or implementation deficiencies. Any breaking change to a stable extension will be announced well in advance of the change being made, with a migration path provided if possible. For an extension to be classified as `Stable` it must be considered `Stable` when used in combination of all possible, non-mutually exclusive, extensions that are already `Stable`. This ensures any extension that is `Stable` will be have the same stability properties when combined with any other extension in the set, while allowing for mutually exclusive extensions to be added. Ideally, no breaking change will be made to a `Stable` extension, with incompatible changes resulting instead in a new, related extension, that is possibly mutually exclusive with the existing one, to enable smooth migration. Some extensions in this category might be `MultiWayIf`, `MonoLocalBinds`, and `ViewPatterns`.

Reviewer 5: * New para (within the same bullet) before "For an extension..."
* New para (within the same bullet) before "Ideally..."
* New para (within the same bullet) before "Some extensions..."

Each has a separate point to make

Reviewer 3: I made this change, making them each new sub-bullets so that it matched the sections for the other categories. Thank you so much!

   * When an extension is deprecated, the user's guide and the compiler warning _must_ include a statement about the longevity of the extension, though this need not necessarily commit to a concrete time for removal. Additionally, the user's guide and the warning should direct users to information about migrating away from the deprecated extension.
   * Any extension will be deprecated prior to removal.
   * Some extensions in this category might be `OverlappingInstances`, `Rank2Types`, `RecordPuns`, and `TypeInType`.

 * `Legacy` extensions are explicitly not recommended for new code, much like `Deprecated`, however, they are expected to be supported indefinitely. This may be to maintain an older language edition, maintain a bridge to a newer feature, or otherwise keep backward compatibility. Extensions that are `Legacy` may have some conflict with `Stable` extensions and explicitly do not need to be considered for an extension to be `Stable`. Some extensions in this category might be `CUSKs`, `DeepSubsumption`,  and `NPlusKPatterns`.

> Above we list expected categorizations of several extensions based on statements in the user's guide. We do not attempt to prescribe any particular categorizations as part of this proposal and simply provide a small number of expectations from the perspective of someone reading the existing documentation.

Reviewer 5: I'd say "several specific extensions"

Reviewer 3: Made this change as well.

> 

Reviewer 5: Start a new subection here: "Lifecycle: how an extension's category may change over time"

(And rename 2.1 to just "Categories"

Reviewer 3: Done, thank you!

The following transitions are included:
 * (does not exist) -> `Experimental`
 * `Experimental` -> `Stable`
 * `Experimental` -> `Deprecated`
 * `Experimental` -> `Legacy`
 * `Stable` -> `Deprecated`
 * `Stable` -> `Legacy`
 * `Legacy` -> `Deprecated`
 * `Deprecated` -> (does not exist)

These transitions are explicitly excluded:
 * (does not exist) -> `Deprecated`

>  * (does not exist) -> `Legacy`

Reviewer 4: I'm not convinced about excluding these. Something we have been known to do in the past is to introduce a new extension name for existing functionality (of the base language or another extension) that is being deprecated or otherwise restricted. For example:
 * `DeepSubsumption` was arguably `Legacy` when it was introduced, because its functionality was already present in previous releases but we wanted to move towards a world in which `NoDeepSubsumption` became the default. (I'm not sure whether we would have considered `NoDeepSubsumption` to be `Experimental` or `Stable` on its introduction.)
 * If we want to partially deprecate an extension, I think the right way is to split it into multiple extensions and introduce one or more of them as `Deprecated` (e.g. I can imagine splitting up `IncoherentInstances = IncoherentInstanceDeclarations + IncoherentClassRoles + IncoherentSolving` with `IncoherentInstanceDeclarations` being `Deprecated`).
 
So I'd suggest moving these to the "discretionary" category.

Reviewer 3: > I'm not convinced about excluding these. Something we have been known to do in the past is to introduce a new extension name for existing functionality (of the base language or another extension) that is being deprecated or otherwise restricted. For example:
> 
> * `DeepSubsumption` was arguably `Legacy` when it was introduced, because its functionality was already present in previous releases but we wanted to move towards a world in which `NoDeepSubsumption` became the default. (I'm not sure whether we would have considered `NoDeepSubsumption` to be `Experimental` or `Stable` on its introduction.)

This is a really good point.

> * If we want to partially deprecate an extension, I think the right way is to split it into multiple extensions and introduce one or more of them as `Deprecated` (e.g. I can imagine splitting up `IncoherentInstances = IncoherentInstanceDeclarations + IncoherentClassRoles + IncoherentSolving` with `IncoherentInstanceDeclarations` being `Deprecated`).

I would have expected all of them to be `Experimental` at first. That way if the way `IncoherentInstances` was cut needed to change then it is less problematic as the entire set was `Experimental`. To be more concrete, if it was realized after being "in the wild" that it is better defined as `IncoherentInstances = IncoherentDeclarationsAndClassRoles + IncoherentSolving` or maybe `IncoherentInstances = IncoherentInstanceDeclarations + IncoherentClassRolesAndSolving` then the only removals/changes would have been made to `Experimental` extensions.

That said, I'm fine with moving this to the "discretionary" category.

This is a small restriction, such that `Stable` or `Legacy` extensions require a deprecation cycle prior to removal.

> The following transitions are possible but left at the discretion of implementors as to be included or not:

Reviewer 3: The intent is to lay out the explicitly acceptable transitions and expected usual lifecycle, the explicitly unacceptable transitions and then those that the proposal is making no judgement on. That is an attempt to balance the communication of an expected lifecycle with some guarantees for predictability and maintaining some flexibility for the implementation.

I'll attempt to adjust the language of this sentence and the following paragraph to better reflect this.

 * (does not exist) -> `Deprecated`
 * (does not exist) -> `Legacy`
 * (does not exist) -> `Stable`
 * `Experimental` -> (does not exist)
 * `Stable` -> `Experimental`
 * `Legacy` -> `Experimental`
 * `Legacy` -> `Stable`
 * `Deprecated` -> `Experimental`
 * `Deprecated` -> `Stable`
 * `Deprecated` -> `Legacy`

> However, it also seems plausible that new knowledge might from time to time cause a stable extension to once again be considered experimental, e.g. in the face of soundness bugs or subtle interactions with other features. We explicitly allow for more transitions to allow for unforeseen circumstances. Restrictions are places on removal only. With this we aim to balance predictibility and flexibility. The consideration of both is important as extension classifications are first and foremost a communications channel between language designers, compiler developers, and users.

Reviewer 7: ```suggestion
However, it also seems plausible that new knowledge might from time to time cause a stable extension to once again be considered experimental, e.g. in the face of soundness bugs or subtle interactions with other features. We explicitly allow for more transitions to allow for unforeseen circumstances. Restrictions are placed on removal only. With this we aim to balance predictibility and flexibility. The consideration of both is important as extension classifications are first and foremost a communications channel between language designers, compiler developers, and users.
```

Reviewer 3: Thanks again, this one should also be fixed now.

For existing, or future, language editions such as `GHC2021` or `Haskell98`, it is expected that none of the contained extensions would be `Experimental`. However, this proposal does not seek to impose any particular policy on the inclusion of extensions into language editions - the developers and the steering committee are always in the best position to make a decision about a concrete extension and extension set.

Moving an extension through the lifecycle will require a GHC proposal. This will give users a chance to provide their input, it will help catch unexpected interactions (such as an `Stable` extension implying an `Experimental` one, and thus potentially allowing breakage at any  moment), and it provides a way for language implementers and users to clarify their mutual expectations.

### 2.2. Documentation

> The user's guide for each extension documents where it is in the extension lifecycle. We expect this to be next to the "Since" and "Implied by" fields in extension documentation. Additionally, the user's guide should provide a history of the extension's sojourn through the various states, and the GHC versions in which it transitioned, for example from being experimental to stable, or from stable to deprecated.

Reviewer 5: The user's guide for each extension should state what category the extension belongs to.  We expect this statement to be next to...

Reviewer 3: Made this change, thank you!

## 3. Effect and Interactions

The tension between large-scale industrial users of Haskell who desire stability and predictability and GHC's traditional role as a venue for language experiments is very real. This system addresses some of the industrial concern by providing clear communication to users that sets realistic expectations for language extensions.

Because this proposal does not change the semantics of any programs, we expect that it will not be a major disturbance.

## 4. Costs and Drawbacks

### 4.1. Development Costs

Development and ongoing costs are expected to exist primarily in the form of additional time to sufficiently warn users of upcoming changes, as well as creating and updating the status of extensions. We expect that the costs associated with moving an extension from one category to another will be small, but the discussion around doing so might take time - in particular, this may result in an increased workload for the GHC Steering Committee to process lifecycle transitions.

## 5. Alternatives

### 5.1. Include Another Categorization

> Another categorization has been discussed at length to provide additional gradient for stability. An additional categorization could allow to further refine the notion of stability. Similarly,  we could categorize in many different ways. However, adding any additional categories is left as a possible enhancement for the future, should a need or desire arise..

Reviewer 5: This para contains virtually no information.  It would be far better to include a link to the discussion. Or just omit.

> Another categorization has been discussed at length to provide additional gradient for stability. An additional categorization could allow to further refine the notion of stability. Similarly,  we could categorize in many different ways. However, adding any additional categories is left as a possible enhancement for the future, should a need or desire arise..

Reviewer 8: ```suggestion
Another categorization has been discussed at length to provide additional gradient for stability. An additional categorization could allow to further refine the notion of stability. Similarly,  we could categorize in many different ways. However, adding any additional categories is left as a possible enhancement for the future, should a need or desire arise.
```

Reviewer 3: Again thank you for the attention to detail here, this should be addressed now.

### 5.2. Adding Compiler Warnings

> One alternative design is to add warnings to the compiler, instead of communicating about the status of extensions only in the User's Guide. We consider this to be likely to provide value, because users typically do not return to sections of the User's Guide that cover features that they already understand. The migration of an extension into a warning set provides a reason for them to consult the documentation. This is left as future work to be covered in other extensions.

Reviewer 5: what sort of warnings?  Again this para is tantalising. It suggests an alternative but does not describe it.  Link?

Reviewer 3: It was in a previous revision of the proposal. Added some wording reflecting that. Not sure if a link to a previous revision is acceptable. If it is, I'm happy to add that or a link to some of the discussion around it that happened previously on this proposal.

> One alternative design is to add warnings to the compiler, instead of communicating about the status of extensions only in the User's Guide. We consider this to be likely to provide value, because users typically do not return to sections of the User's Guide that cover features that they already understand. The migration of an extension into a warning set provides a reason for them to consult the documentation. This is left as future work to be covered in other extensions.

Reviewer 7: ```suggestion
One alternative design is to add warnings to the compiler, instead of communicating about the status of extensions only in the User's Guide. We consider this to be likely to provide value, because users typically do not return to sections of the User's Guide that cover features that they already understand. The migration of an extension into a warning set provides a reason for them to consult the documentation. This is left as future work to be covered in other proposals.
```

Reviewer 3: Thank you! Should be fixed now.

  * Whether they change the meaning of a program when enabled (such as `ScopedTypeVariables` and `MonoLocalBinds`) and those that add features without changing existing programs (such as `MultiWayIf`, `NPlusKPatterns`, and `LinearTypes`)
  * Extensions that merely enable syntactic sugar (such as `MultiWayIf`, `NPlusKPatterns`, `DoAndIfThenElse`) and those that enrich the type system with new features (such as `DataKinds`, `TypeFamilies`, or `PolyKinds`).

This is clearly something that's desirable for users to know. However, this information is orthogonal to the stability, so we believe that mechanisms to communicate these aspects should be designed in a separate proposal.

### 6. Related Work

There are several other proposals and discussions that are related. There is a general [discussion thread](https://github.com/ghc-proposals/ghc-proposals/discussions/635) around features that are stable or not. Also a much more broadly sweeping [proposal](https://github.com/ghc-proposals/ghc-proposals/pull/636) includes a notion of stability. This proposal is now aiming to be very targeted in effort to make progress.

## 7. Unresolved Questions

## 8. Implementation Plan

> The Haskell Foundation will either recruit volunteers or pay for the technical work of implementing this proposal. The classifications of individual language extensions will be left to a future proposal, and those discussions can proceed while implementation is ongoing, should this proposal be accepted.

Reviewer 4: I think this paragraph comes from an earlier revision of the proposal, in which there were more warnings being introduced. I don't think the current version has anything to implement until the classification is done.

> The Haskell Foundation will either recruit volunteers or pay for the technical work of implementing this proposal. The classifications of individual language extensions will be left to a future proposal, and those discussions can proceed while implementation is ongoing, should this proposal be accepted.

Reviewer 3: Yes that is a great point. I'll adjust the language to make it clear there is nothing to implement at this proposal.

Note that this warning syntax rules out the names `Deprecated`, `Stable`, `Experimental`, and `Legacy` for future language extensions.

#### 2.2.1. Warnings are for Enabled Extensions

Extension warnings will be issued when an extension is enabled, not when the language features that it supports are used in a program. The following program should issue a warning, even though no multi-way conditionals occur in it:
```
{-# OPTIONS_GHC -WXMultiWayIf #-}
{-# LANGUAGE MultiWayIf #-}
module Main where

main :: IO ()
main = pure ()
```

There are two main reasons to do this, rather than wait for the associated code path in the compiler to be invoked:

 1. Users don't need to think about when one extension implies features from another. The mental model for warnings is simple and straightforward, and it's easier to see the connection between a change in the code and a new warning in CI.
 
 2. The implementation is likely to be simpler, because the extension lifecycle warnings can presumably be issued without needing to scatter them throughout the parser and type checker.

> Warnings should be issued for extensions no matter how they are enabled. For instance, the following program should also issue a warning:

Reviewer: What location is used for the warning? It is clear where to put it when there is a language pragma, but not so here. And presumably we don't want to emit a warning for every module that GHC compiles in this case? Or do we? Perhaps it should be like `-Wunused-packages`, and not have a location.

Author: Should the specific source locations of warnings belong here? I saw this as something like the specific wordings - up to the one doing the implementation.

Off the cuff, I'd think that the best possibility would be to put as many as possible on `LANGUAGE` or `OPTIONS_GHC` pragmas, but when that can't happen, either do it locationless or on the `module` header at the top if there is one? But this seems like the kind of thing that the implementers should decide, and adjust as needed.

Reviewer 4: I agree we don't need to pin down the implementation too heavily regarding source locations. But it's a good question as to whether warnings are issued once per module or not. This section as written implies they are, so e.g. adding an experimental extension to `default-extensions` would result in a warning on every module in the program. Is that what we want, when arguably the location we'd like to point the user to is in the `.cabal` file?

It is pretty clear that we should report a per-module warning on any `LANGUAGE` or `OPTIONS_GHC` pragma that enables a warned-about extension. But for extensions that are not explicitly enabled in the file, rather than each module reporting duplicate warnings for extensions that were enabled globally, perhaps we should have the driver report them once?

I'm in two minds about this. One consequence would be that in the following example, we wouldn't get a warning, because `-WXBangPatterns` is local to the file whereas `BangPatterns` was enabled globally:
```hs
{-# OPTIONS_GHC -WXBangPatterns #-}
module Main where
```

Reviewer: I guess it seems to me that the location is part of the UX of this feature, and we _have_ been discussing UX a bit in this proposal.

Reviewer 4: Thanks for the detailed response @david-christiansen, it's fairly convincing that one-module-at-a-time is a reasonable way forward.

> The second difficulty is that if the `.cabal` file turns on an experimental extension, but each module has a pragma that turns it off, then the extension is in fact not enabled anywhere. As a user, I'd expect the warning to not occur until I create a file without the pragma that turns it off.

What about `ghci`/`cabal repl`? Presumably it should emit warnings about experimental extensions, even in the absence of any modules?

Reviewer: > What about ghci/cabal repl? Presumably it should emit warnings about experimental extensions, even in the absence of any modules?

In fact, I think there are in principle a whole class of awkward cases (although this is getting quite far off-topic). Suppose that, for example, the GHC team wanted to remove a flag from GHC. Arguably, it would be nice to emit a deprecation warning _about the flag_ for a release or two before removing it. But this has exactly the same problems that we've been discussing (since setting extensions is kind of a subset of setting a flag), but it's harder to avoid the "no location" case.

Thinking of it this way does make me think slightly differently about some of David's analysis.

> The second difficulty is that if the .cabal file turns on an experimental extension, but each module has a pragma that turns it off, then the extension is in fact not enabled anywhere. As a user, I'd expect the warning to not occur until I create a file without the pragma that turns it off.

If we consider the case of deprecated extensions, _this is still problematic_! Because if the extension gets removed, your build will still break because GHC will reject the flag despite it never being removed. So that makes me think that you _should_ get a warning here.

This makes me suggest a slightly extended version of the simple approach that David suggested, but this time extending to _anything that sets a flag_, extension or otherwise.
- A warning is emitted _when the flag is enabled_, regardless of whether it is later disabled or whether it will not take effect for some other reason. The "warnable event" is turning on the flag, that's it (easy to understand!)
- The warning is emitted where the flag is set, if possible:
    - For pragmas it's easy
    - For command line flags (or e.g. `:set` in GHCi), we emit a no-location warning

This doesn't have the best UX, but it's mechanically relatively simple (I think) and easy for users to understand why it's doing what it's doing in the end.

Reviewer 2: > A warning is emitted when the flag is enabled, regardless of whether it is later disabled 

Won't this be awkward in the case of `-XGHC20xy -XNoDeprecatedExtensionPartOfGHC20xy`?

Author: I think that Joachim's concern here should be taken seriously. Enabling a set and then disabling a few elements of it is a common practice. We shouldn't train people to turn off the warning by giving them false positives.

How about this as a loose specification: a warning is emitted if GHC ever enters a state where it will accept code when the warnable extension is enabled? These states are, so far as I can see, when a prompt becomes available at GHCi or when a module header has been parsed.

This seems to cover language pragmas, options pragmas, command-line flags, GHCi, and all the other situations I can think of. At the same time, it's short and uncomplicated, and doesn't deliver false positives.

Reviewer: > Won't this be awkward in the case of -XGHC20xy -XNoDeprecatedExtensionPartOfGHC20xy?

If a part of GHC20xy is deprecated, then it is going to be removed in the future, perhaps breaking GHC20xy entirely? In this situation it seems like maybe I should be worried.

Personally I'd be willing to bite this bullet but I can see why you might not want to.

Reviewer 2: Maybe I should have said `-XNoLegacyExtensionButPartOfGHC20xy`.

Or a setting `-WXBenignExtensionButUnwantedInThisOrg` followed by `{-# LANGUAGE GHC2023 #-}{-# LANGUAGE NoBenignExtensionButUnwantedInThisOrg #-}`

Reviewer: I guess the thing I'm gesturing at is that for _removal_, the thing that is at risk of breakage is not just your _code_ but also your _flag specification_. If the extension `Foo` is removed, then my build which says `-XFoo -XNoFoo` _will break_.

Maybe we're fine with that! Perhaps the scope of this is really just the source code. After all, we don't currently tend to offer deprecation periods or warnings for flags that get removed. But maybe we should!

Author: I think that this suggests something that's out of scope of this proposal, but that would be a very sensible policy: if an extension is removed, disabling it should be a no-op rather than an error. In other words, lets say for the sake of argument that we remove the extension `AlternativeLayoutRuleTransitional` in the future. Code that specifies `{#- LANGUAGE NoAlternativeLayoutRuleTransitional #-}` should continue to work indefinitely, as should code that is built with `-XAlternativeLayoutRuleTransitional` but then has that language pragma in the header to turn it off. The "invalid extension" error for extensions that have been removed should only occur when GHC is in a state where it will accept Haskell code, and when the extension is in fact enabled.

I generally think that backwards-incompatible changes shouldn't be happening, but there's always a cost to keeping things around, and it seems plausible that a language extension will be removed at some point in the future, after providing a good migration period and fixing up Hackage.

This means of handling formerly-valid extensions should be possible to build in one place, test thoroughly, and document well, and it future-proofs the process as well as possible. It also won't break the builds in the way @michaelpj describes.

Right?

Reviewer 4: This all sounds good to me, and I think could in fact be specified in this proposal as a new lifecycle category: an extension can be `PiningForTheFjords`, at which point GHC will allow it to be disabled as you describe. And there's an expected transition `Deprecated` -> `PiningForTheFjords`.

Although perhaps it would be enough to make an unrecognised extension a warning rather than an error. For some reason `-XScopedWombatPunning` is currently a fatal error, whereas an unrecognised warning flag such as `-Wscoped-wombat-puns` merely issues a warning.

Reviewer 4: > Although perhaps it would be enough to make an unrecognised extension a warning rather than an error.

FWIW I posted a GHC ticket proposing this behaviour: https://gitlab.haskell.org/ghc/ghc/-/issues/23858

Author: I've gone back and forth about a "removed" category for this proposal, as you suggest here. It has some nice benefits:
 1. It provides a natural way to have the User's Guide describe why some old program no longer works (the feature was removed in version Y, for this reason, here's how to migrate)
 2. It provides a good way for the compiler to direct users to replacement features or to said documentation page
 3. It allows typos in language feature names to be more than warnings :-)

On the other hand, it seems that the language of pragmas, feature flags and other configurations that GHC accepts has a more general problem with breaking changes, and that language extension flags are somehow a special case of that problem. I've never been a power user of GHC flags, so I'm a poor person to propose a specific plan, but some kind of migration/lifecycle system for compiler flags and options seems useful and important.

I know I'm entering the typical Haskell-culture failure mode here of rejecting a solution that works in favor of a nonexistent theoretical one that could be better in principle, so I'm definitely willing to revise this conclusion.

#### 2.2.4. Identifying Flags

When possible, the implementation should distinguish between extension lifecycle warnings that result from extension set flags and those that result from individual flags. This is because the individual extension warnings may be used for reasons other than stability - a company might decide to warn on an extension simply because the team has decided that it doesn't fit the house style.

### 2.3. Documentation

The user's guide for each extension documents where it is in the extension lifecycle. We expect this to be next to the "Since" and "Implied by" fields in extension documentation. Additionally, the user's guide should provide a history of the extension's sojourn through the various states, and the GHC versions in which it went from being experimental to stable, or from stable to deprecated.

## 3. Examples

With the use of `-Wcompat`, and presuming the categorization of `IncoherentInstances` as `Deprecated`, matching the statements in the user's guide, consider the following code snippet:
```haskell
{-# LANGUAGE IncoherentInstances #-}
module Foo where
```

> This would result in a warning at compilation time stating the use of a `Deprecated` extension.

Reviewer: ```suggestion
This would result in a warning on the language pragma at compilation time stating the use of a `Deprecated` extension.
```

## 8. Implementation Plan

> The classifications of individual language extensions will be left to a future proposal, and those discussions can proceed while implementation is ongoing, should this proposal be accepted. Thus there is nothing concrete to implement as part of this proposal.

Reviewer 3: Absolutely! Made that exact change.
